### PR TITLE
Add additional game features and actions

### PR DESCRIPTION
**Pull request by TheEmidee**
Create struct holding additional feaures and actions.
Create map with that struct and a string for options.
Create functions to get all the features and actions.
Update getting all features and actions by calling the new functions.

### CODE REVIEW

File: Source/GameBaseFramework/Private/GameFramework/Experiences/GBFExperienceDefinition.cpp [hunk 1]
```diff
@@ -15,6 +18,78 @@ FPrimaryAssetType UGBFExperienceDefinition::GetPrimaryAssetType()
     return PrimaryAssetType;
 }
 
+TArray< FString > UGBFExperienceDefinition::GetAllGameFeatures( const UWorld * world ) const
+{
+    auto result = GameFeaturesToEnable;
```
Comment (TheEmidee) [inline Source/GameBaseFramework/Private/GameFramework/Experiences/GBFExperienceDefinition.cpp]: This creates a copy. Turns out it was better to fill the arrays by reference.
In addition to doing that:
* Use AddRange to add the whole `GameFeaturesToEnable` at once
* Reserve enough items in the array before passing it to this function (like twice the number of game features)

Do the same for actions

File: Source/GameBaseFramework/Private/GameFramework/Experiences/GBFExperienceManagerComponent.cpp [hunk 2]
```diff
@@ -83,14 +81,16 @@ void UGBFExperienceManagerComponent::EndPlay( const EEndPlayReason::Type EndPlay
             }
         };
 
-        deactivate_list_of_actions( CurrentExperience->Actions );
-        for ( const auto * action_set : CurrentExperience->ActionSets )
+        for ( auto index = LoadedGameFeatureActions.Num() - 1; index >= 0; --index )
         {
-            if ( action_set != nullptr )
+            auto * action = LoadedGameFeatureActions[ index ];
```
Comment (TheEmidee) [inline Source/GameBaseFramework/Private/GameFramework/Experiences/GBFExperienceManagerComponent.cpp]: if ( auto * action = ... )

File: Source/GameBaseFramework/Private/GameFramework/Experiences/GBFExperienceManagerComponent.cpp [hunk 3]
```diff
@@ -289,40 +289,31 @@ void UGBFExperienceManagerComponent::OnExperienceLoadComplete()
...
-            if ( UGameFeaturesSubsystem::Get().GetPluginURLForBuiltInPluginByName( plugin_name, /*out*/ plugin_url ) )
-            {
-                This->GameFeaturePluginURLs.AddUnique( plugin_url );
-            }
-            else
-            {
-                ensureMsgf( false, TEXT( "OnExperienceLoadComplete failed to find plugin URL from PluginName %s for experience %s - fix data, ignoring for this run" ), *plugin_name, *context->GetPrimaryAssetId().ToString() );
-            }
-        }
+    const TArray< FString > game_features = CurrentExperience->GetAllGameFeatures( GetWorld() );
```
Comment (TheEmidee) [inline Source/GameBaseFramework/Private/GameFramework/Experiences/GBFExperienceManagerComponent.cpp]: const auto

File: Source/GameBaseFramework/Public/GameFramework/Experiences/GBFExperienceDefinition.h [hunk 4]
```diff
@@ -44,4 +56,7 @@ class GAMEBASEFRAMEWORK_API UGBFExperienceDefinition final : public UPrimaryData
     // List of additional action sets to compose into this experience
     UPROPERTY( EditDefaultsOnly, Category = Gameplay )
     TArray< UGBFExperienceActionSet * > ActionSets;
+
+    UPROPERTY( EditDefaultsOnly, Category = Gameplay )
+    TMap< FString, FGBFExperienceAdditionalActionSets > OptionToAdditionalActionSetsMap;
```
Comment (TheEmidee) [inline Source/GameBaseFramework/Public/GameFramework/Experiences/GBFExperienceDefinition.h]: I'd rather have a `TArray< FGBFExperienceAdditionalActionSets >` here, and move the FString in the structure

File: Source/GameBaseFramework/Public/GameFramework/Experiences/GBFExperienceManagerComponent.h [hunk 5]
```diff
@@ -77,6 +78,7 @@ class GAMEBASEFRAMEWORK_API UGBFExperienceManagerComponent final : public UGameS
 
     int32 NumGameFeaturePluginsLoading = 0;
     TArray< FString > GameFeaturePluginURLs;
+    TArray< UGameFeatureAction * > LoadedGameFeatureActions;
```
Comment (TheEmidee) [inline Source/GameBaseFramework/Public/GameFramework/Experiences/GBFExperienceManagerComponent.h:81]: UPROPERTY()

File: Source/GameBaseFramework/Private/GameFramework/Experiences/GBFExperienceManagerComponent.cpp (hunk 3 above)
Comment (Frickerson) [reply Source/GameBaseFramework/Private/GameFramework/Experiences/GBFExperienceManagerComponent.cpp]: Not valid anymore with newest changes.

File: Source/GameBaseFramework/Private/GameFramework/Experiences/GBFExperienceManagerComponent.cpp (hunk 2 above)
Comment (Frickerson) [reply Source/GameBaseFramework/Private/GameFramework/Experiences/GBFExperienceManagerComponent.cpp]: Done.

File: Source/GameBaseFramework/Private/GameFramework/Experiences/GBFExperienceDefinition.cpp (hunk 1 above)
Comment (Frickerson) [reply Source/GameBaseFramework/Private/GameFramework/Experiences/GBFExperienceDefinition.cpp]: Done.

File: Source/GameBaseFramework/Private/GameFramework/Experiences/GBFExperienceManagerComponent.cpp [hunk 6]
```diff
@@ -47,9 +45,9 @@ void UGBFExperienceManagerComponent::EndPlay( const EEndPlayReason::Type EndPlay
     Super::EndPlay( EndPlayReason );
 
     // deactivate any features this experience loaded
-    //@TODO: This should be handled FILO as well
-    for ( const auto & plugin_url : GameFeaturePluginURLs )
+    for ( auto index = GameFeaturePluginURLs.Num() - 1; index >= 0; --index )
     {
+        auto plugin_url = GameFeaturePluginURLs[ index ];
```
Comment (TheEmidee) [inline Source/GameBaseFramework/Private/GameFramework/Experiences/GBFExperienceManagerComponent.cpp]: const auto &

Comment (Frickerson) [reply Source/GameBaseFramework/Private/GameFramework/Experiences/GBFExperienceManagerComponent.cpp]: Done.

**Pull request merged**